Refactor language worker layer for HttpInvoker - Part 1

pragnagopa opened the pull request:
Part of #4887 
This PR addresses:
- Code refactoring to allow code reuse between Rpc and http invocation paths
- Added language agnostic HttpFunctionDescriptorProvider 
- Added HttpInvokerOptions

Fixes #4919  Fixes #4976 Fixes #4929 and Fixes #4972

pragnagopa commented:
@fabiocav - Do you recommend working on a branch until I have all the changes wired up ?

Note: This PR refactoring only. Will setup a sync to go over the design.

pragnagopa commented:
Ping! Making sure this is on your radar for review.

yojagad commented on src/WebJobs.Script/OutOfProc/Http/HttpInvokerChannelFactory.cs:
```diff
@@ -0,0 +1,56 @@
+            {
+                throw new InvalidOperationException($"Found more than one HttpInvoker config");
+            }
+            string workerId = Guid.NewGuid().ToString();
```
>string workerId = Guid.NewGuid().ToString(); [](start = 12, length = 44)

Would it be useful to augment workerId with "_httpInvoker" and "_workerInvoker" so that we see in logs or does the fact that CreateLogger($"Worker.HttpInvokerChannel.{workerId}"); below already cover it?

pragnagopa replied:
Yes. Using Worker.HttpInvokerChannel will differentiate between language worker and http invoker

fabiocav commented on src/WebJobs.Script/Eventing/EventSources.cs:
```diff
@@ -1,12 +1,6 @@
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
-using System;
-using System.Collections.Generic;
```
nit: keep default using statements.

fabiocav commented on src/WebJobs.Script/Host/FunctionMetadataManager.cs:
```diff
@@ -276,4 +276,4 @@ internal static string ParseLanguage(string scriptFilePath, IEnumerable<WorkerCo
             return null;
         }
     }
-}
```
nit: undo whitespace change in this PR (since this PR is already pretty big, let's scope to the required changes)

pragnagopa replied:
missed to include a change in this file. Next iteration will handle that.

fabiocav commented on src/WebJobs.Script.Grpc/Abstractions/WorkerDescription.cs:
```diff
@@ -57,4 +57,4 @@ public string GetWorkerPath()
             }
         }
     }
-}
```
nit: undo whitespace change in this PR (since this PR is already pretty big, let's scope to the required changes)

pragnagopa replied:
done!

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/MessageExtensions/HttpMessageConversionExtensions.cs:
```diff
@@ -0,0 +1,56 @@
+            return HttpMessageExtensionUtilities.ConvertFromHttpMessageToExpando(responseMessage);
+        }
+
+        public static string ToHttpHeader(this object value)
```
Let's avoid an extension method on `object`

pragnagopa replied:
done.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/MessageExtensions/HttpMessageConversionExtensions.cs:
```diff
@@ -0,0 +1,56 @@
+    {
+        private static readonly JsonSerializerSettings _datetimeSerializerSettings = new JsonSerializerSettings { DateParseHandling = DateParseHandling.None };
+
+        public static Task<object> ToObject(this HttpResponseMessage responseMessage)
```
This is essentially exposing a new API for the same functionality supported by the `HttpMessageExtensionUtilities`. Let's either use that from the caller or move the functionality in `HttpMessageExtensionUtilities` to `HttpResponseExtensions`

pragnagopa replied:
yes. I still have to clean up the converters.

pragnagopa replied:
deleted these. Will add converters in a separate PR.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/MessageExtensions/HttpMessageConversionExtensions.cs:
```diff
@@ -0,0 +1,56 @@
+            return value.ToString();
+        }
+
+        private static string GetStringRepresentationOfBody(HttpRequest request)
```
Was the intention to have this be an extension method? If so, we should move to `HttpRequestExtensions`

pragnagopa replied:
do you recommend static methods instead of extensions? the purpose of these helpers is to convert from input binding data to Http Request post body

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/MessageExtensions/HttpMessageConversionExtensions.cs:
```diff
@@ -0,0 +1,56 @@
+            {
+                result = reader.ReadToEnd();
+            }
+            // request.Body.Position = 0;
```
Commented out code. Was this really not needed?

pragnagopa replied:
Apologies. I haven't up cleaned code for converters yet. I have separate issue to address it.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/MessageExtensions/HttpMessageExtensionUtilities.cs:
```diff
@@ -0,0 +1,34 @@
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Script.OopHttp
```
Why the different namespace here?

pragnagopa replied:
fixed namespace! Same here the purpose of this converter is to convert from HttpResponse message to http output binding.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/MessageExtensions/HttpMessageExtensionUtilities.cs:
```diff
@@ -0,0 +1,34 @@
+            {
+                expando.body = await inputMessage.Content.ReadAsStringAsync();
+            }
+            //return expando;
```
Commented out code. The actual implementation is returning the body only, doesn't look right based on the additional logic above.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/DefaultHttpInvokerService.cs:
```diff
@@ -0,0 +1,31 @@
+        {
+            var requestUri = _httpInvokerBaseUrl + scriptInvocationContext.FunctionMetadata.Name;
+            // TODO: convert send ScriptInvocationContext to http request and send to http invoker
+            HttpContent content = new StringContent("hello");
```
can we remove the extra code here and leave just the exception?

pragnagopa replied:
yes. removed implementation of this service. will send a seperate PR

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+
+        public bool IsSupported(FunctionMetadata functionMetadata, string workerRuntime)
+        {
+            return true;
```
Is there a "to do" here?

pragnagopa replied:
good catch! thanks. Runtime check is not needed in the dispatcher. Cleaned up interface

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+        {
+            _functions = functions;
+
+            if (functions == null || functions.Count() == 0)
```
change `functions.Count() == 0` to `functions.Any()`

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+
+        public async void WorkerError(HttpWorkerErrorEvent workerError)
+        {
+            if (!_disposing)
```
Is this error handling logic copied from the OOP dispatcher?

fabiocav replied:
What can trigger worker errors? Will any invocation that fails to communicate with the worker end up here?

pragnagopa replied:
this is the process exit error handler. if there is no  worker process that can handle invocation and retries to start the process fail, functions host is shutdown

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+
+        private async Task DisposeAndRestartWorkerChannel(string workerId)
+        {
+                _logger.LogDebug("Disposing channel for workerId: {channelId}", workerId);
```
Will this have an issue with multiple concurrent callers? Do we need to synchronize this logic?

pragnagopa replied:
this is gets triggered by the process exit handler. And for a give process, it only triggers once per worker process

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+        {
+            if (!_disposed && disposing)
+            {
+                _logger.LogDebug("Disposing HttpFunctionInvokeDispatcher");
```
nit: use `nameof` for the type name

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+
+namespace Microsoft.Azure.WebJobs.Script.OutOfProc
+{
+    internal class HttpFunctionInvokeDispatcher : IFunctionDispatcher
```
nit: type name - `HttpFunctionInvocationDispatcher` maybe?

pragnagopa replied:
👍

yojagad commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvokeDispatcher.cs:
```diff
@@ -0,0 +1,200 @@
+
+        public FunctionDispatcherState State { get; private set; }
+
+        internal ConcurrentStack<HttpWorkerErrorEvent> LanguageWorkerErrors => _languageWorkerErrors;
```
>LanguageWorkerErrors [](start = 55, length = 20)

Should these be called httpWorkerErrors or something? applies to JobHostLanguageWorkerChannel too.

pragnagopa replied:
done. will clean up

pragnagopa commented on src/WebJobs.Script.Grpc/Abstractions/WorkerDescriptionExtensions.cs:
```diff
@@ -7,7 +7,7 @@ namespace Microsoft.Azure.WebJobs.Script.Abstractions
     public static class WorkerDescriptionExtensions
     {
-        public static void Validate(this WorkerDescription workerDescription)
+        public static void ValidateRpcWorkerDescription(this WorkerDescription workerDescription)
```
Good idea. I will open a tracking issue: https://github.com/Azure/azure-functions-host/issues/4972

pragnagopa commented on src/WebJobs.Script/Description/OutOfProc/Rpc/RpcFunctionDescriptorProvider.cs:
```diff
@@ -0,0 +1,43 @@
+{
+    internal class RpcFunctionDescriptorProvider : OutOfProcDescriptorProviderBase
+    {
+        private readonly ILoggerFactory _loggerFactory;
```
done!

pragnagopa commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -54,6 +54,11 @@ public static bool IsLegacyPlaceholderTemplateSite(this IEnvironment environment
             return string.IsNullOrEmpty(siteName) ? false : siteName.Equals(ScriptConstants.LegacyPlaceholderTemplateSiteName, StringComparison.InvariantCultureIgnoreCase);
         }
 
+        public static bool IsHttpInvokerEnabled(this IEnvironment environment)
```
correct. temporary for me test the wire up. tracking issue https://github.com/Azure/azure-functions-host/issues/4929

pragnagopa commented on src/WebJobs.Script/Eventing/OutOfProc/Http/HttpWorkerRestartEvent.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class HttpWorkerRestartEvent : ScriptEvent
+    {
+        internal HttpWorkerRestartEvent(string workerId)
```
thanks. not intentional. fixing it

pragnagopa commented on src/WebJobs.Script/Description/OutOfProc/OutOfProcDescriptorProviderBase.cs:
```diff
@@ -16,19 +16,17 @@
             _dispatcher = dispatcher;
             _loggerFactory = loggerFactory;
-            _workerRuntime = workerRuntime;
         }
```
done!

pragnagopa commented:
Thanks for the feedback. Addressed most of the the comments. Please note:
- I still have to clean up Http converters
- I am using environment variable to test the http invoker pipeline temporarily.
Assuming refactored folders and names are looking ok. I will clean up the commits and turn it into a PR

pragnagopa commented:
Removed http converters. Ready for review! Thanks

mhoeger commented on src/WebJobs.Script.WebHost/Program.cs:
```diff
@@ -67,7 +67,7 @@ public static IWebHostBuilder CreateWebHostBuilder(string[] args = null)
                 {
                     loggingBuilder.ClearProviders();
-
+                    loggingBuilder.AddConsole();
```
Do you mind explaining what this change does?

pragnagopa replied:
good catch...unintenational. This is useful for local debugging to see logs from worker processes.

mhoeger commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -54,6 +54,11 @@ public static bool IsLegacyPlaceholderTemplateSite(this IEnvironment environment
 
+        public static bool IsHttpInvokerEnabled(this IEnvironment environment)
+        {
+            return environment.GetEnvironmentVariable(FunctionsHttpInvoker) == "1";
```
Nit: do you mind adding a TODO with linked work item if this is temporary?

pragnagopa replied:
will do!

pragnagopa replied:
removed in this part 2 PR.

mhoeger commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -0,0 +1,190 @@
+
+namespace Microsoft.Azure.WebJobs.Script.OutOfProc
+{
+    internal class HttpFunctionInvocationDispatcher : IFunctionDispatcher
```
It seems like the process management logic is being duplicated (ex: methods around starting/stopping, HttpWorkerErrorEvent and HttpWorkerRestartEvent), but I'm not sure if it needs to be.

It seems like we should be able to separate out the process management portions from the invocation / setup. Maybe this a change for a future iteration, but I think that it's going to be difficult to maintain two separate code paths for process management, when what needs to be done is largely the same.

pragnagopa replied:
I agree. yes this is one part i did not refactor yet. Opened https://github.com/Azure/azure-functions-host/issues/4977

mhoeger commented on src/WebJobs.Script/OutOfProc/Rpc/Configuration/WorkerConfigFactory.cs:
```diff
@@ -111,37 +119,69 @@ internal void AddProvidersFromAppSettings()
+        internal void AddHttpInvokerProviderFromAppSettings()
+        {
+            var httpInvokerSection = _config.GetSection($"{LanguageWorkerConstants.HttpInvokerSectionName}");
+            var workerDirectorySection = httpInvokerSection.GetSection(LanguageWorkerConstants.WorkerDirectorySectionName);
```
Is this temporary or is the HTTP invoker path going to be configured by another worker.config.json (or worker.config.json-like file)?

pragnagopa replied:
this will be configured on host.json tracking issue https://github.com/Azure/azure-functions-host/issues/4929 . Working on updating this PR to include options setup for http invoker. Apologies for the confusion,

pragnagopa commented:
PR for part 2: Adding httpInvoker section in host.json https://github.com/pragnagopa/azure-functions-host/pull/2
Please review.

pragnagopa commented:
PR for part 3:  Implement HttpInvokerService pragnagopa#3
Please review.

pragnagopa commented:
PR for part 4: Find unused Tcp Port to host http invoker service and pass port number as command line https://github.com/pragnagopa/azure-functions-host/pull/4

fabiocav commented on src/WebJobs.Script/OutOfProc/Rpc/LanguageWorkerProcessFactory.cs:
```diff
@@ -42,7 +42,7 @@ public ILanguageWorkerProcess CreateLanguageWorkerProcess(string workerId, strin
         {
             WorkerConfig workerConfig = _workerConfigs.Where(c => c.Language.Equals(runtime, StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
             ILogger workerProcessLogger = _loggerFactory.CreateLogger($"Worker.LanguageWorkerProcess.{runtime}.{workerId}");
-            return new LanguageWorkerProcess(runtime, workerId, scriptRootPath, _rpcServer.Uri, workerConfig.Arguments, _eventManager, _workerProcessFactory, _processRegistry, workerProcessLogger, _consoleLogSource);
```
Should the type be renamed since it's now creating an instance of an `RpcWorkerProcess`?

fabiocav commented on src/WebJobs.Script/Utility.cs:
```diff
@@ -497,6 +497,19 @@ internal static string GetWorkerRuntime(IEnumerable<FunctionMetadata> functions)
             return null;
         }
 
+        internal static bool IsSupported(FunctionMetadata functionMetadata, string workerRuntime)
```
This needs a better name now that it is in this generic class.

pragnagopa replied:
gave a really long name :)

pragnagopa commented on src/WebJobs.Script/Description/OutOfProc/Rpc/RpcFunctionDescriptorProvider.cs:
```diff
@@ -0,0 +1,41 @@
+        public override async Task<(bool, FunctionDescriptor)> TryCreate(FunctionMetadata functionMetadata)
+        {
+            if (functionMetadata == null)
+            {
```
This needs to be overridded as Rpc descriptor enforces language.

pragnagopa commented on src/WebJobs.Script/OutOfProc/Http/IHttpInvokerService.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public interface IHttpInvokerService
+    {
+        Task GetInvocationResponse(ScriptInvocationContext scriptInvocationContext);
```
This is used to add typed HttpClient. Do you recommend renaming this to: IHttpInvokerClient ?

pragnagopa commented on src/WebJobs.Script/OutOfProc/ProcessManagement/WorkerProcessBase.cs:
```diff
@@ -22,48 +22,36 @@ internal class LanguageWorkerProcess : ILanguageWorkerProcess, IDisposable
 
         internal Queue<string> ProcessStdErrDataQueue => _processStdErrDataQueue;
 
+        internal abstract Process CreateWorkerProcess();
```
we have an existing issue tracking refactoring code that starts a process: https://github.com/Azure/azure-functions-host/issues/3050 . Please let me know if there is anything more specific you were thinking about.

pragnagopa commented:
@fabiocav - Combining Part 2 of the PR: https://github.com/pragnagopa/azure-functions-host/pull/2 with this PR.

pragnagopa commented:
@maiqbal11  - fyi

pragnagopa commented:
@maiqbal11 - Please verify merge changes in workerconfigfactory and RpcWorkerDescription.

@fabiocav - merged with latest changes in dev. Please take a look

maiqbal11 commented:
> @maiqbal11 - Please verify merge changes in workerconfigfactory and RpcWorkerDescription.
> 
> @fabiocav - merged with latest changes in dev. Please take a look

Refactor looks good to me. Should be double good if all tests for this change pass!

fabiocav commented on src/WebJobs.Script/Extensions/HttpRequestExtensions.cs:
```diff
@@ -72,5 +73,25 @@ public static bool IsColdStart(this HttpRequest request)
+            return bytes;
+        }
+
+        public static string GetRequestBodyAsString(this HttpRequest request)
```
Use `ReadAsStringAsync` instead.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -0,0 +1,180 @@
+
+        public FunctionDispatcherState State { get; private set; }
+
+        internal async void InitializeJobhostLanguageWorkerChannelAsync()
```
Doesn't look like this method is used.

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpInvokerChannel.cs:
```diff
@@ -0,0 +1,83 @@
+{
+    public class HttpInvokerChannel : IHttpInvokerChannel, IDisposable
+    {
+        private readonly string _rootScriptPath;
```
`_rootScriptPath` and  `_eventManager` are unused

`_workerId` could likely become an auto property here.

pragnagopa replied:
done!

fabiocav commented on src/WebJobs.Script/OutOfProc/Http/HttpInvokerProcessFactory.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Microsoft.Azure.WebJobs.Script.OutOfProc.Http
+{
+    internal class HttpInvokerProcessFactory : IHttpInvokerProcessFactory
```
`HttpInvokerProcessFactory` -> `HttpWorkerProcessFactory` (including interface)

fabiocav commented on src/WebJobs.Script/OutOfProc/Rpc/LanguageWorkerProcessFactory.cs:
```diff
@@ -38,11 +38,11 @@ internal class LanguageWorkerProcessFactory : ILanguageWorkerProcessFactory
+            WorkerConfig workerConfig = _workerConfigs.Where(c => c.Description.Language.Equals(runtime, StringComparison.OrdinalIgnoreCase)).FirstOrDefault();
             ILogger workerProcessLogger = _loggerFactory.CreateLogger($"Worker.LanguageWorkerProcess.{runtime}.{workerId}");
-            return new LanguageWorkerProcess(runtime, workerId, scriptRootPath, _rpcServer.Uri, workerConfig.Arguments, _eventManager, _workerProcessFactory, _processRegistry, workerProcessLogger, _consoleLogSource);
+            return new RpcWorkerProcess(runtime, workerId, scriptRootPath, _rpcServer.Uri, workerConfig.Arguments, _eventManager, _workerProcessFactory, _processRegistry, workerProcessLogger, _consoleLogSource);
```
Rename the type from `LanguageWorkerProcessFactory` to `RpcWorkerProcessFactory`
